File: apple/LC716MaxStack.java
class MaxStack {
    class Node {
        int value;
        boolean valid;
        int index;
        Node (int v, boolean vl, int i) {
            this.value = v;
            this.valid = vl;
            this.index = i;
        }
    }

    Queue<Node> pq;
    List<Node> list;
    public MaxStack() {
        list = new ArrayList<>();
        pq = new PriorityQueue<>((x1, x2)->(x2.value != x1.value ? x2.value-x1.value : x2.index-x1.index));
    }

    public void push(int x) {
        Node node = new Node(x, true, list.size());
        pq.offer(node);
        list.add(node);
    }

    public int pop() {
        while (list.size()>0 && !list.get(list.size()-1).valid) list.remove(list.size()-1);
        Node node = list.remove(list.size()-1);
        pq.remove(node);
        return node.value;
    }

    public int top() {
        while (list.size()>0 && !list.get(list.size()-1).valid) list.remove(list.size()-1);
        return list.get(list.size()-1).value;
    }

    public int peekMax() {
        return pq.peek().value;
    }

    public int popMax() {
        Node node = pq.poll();
        node.valid = false;
        return node.value;
    }
}

/**
 * Your MaxStack object will be instantiated and called as such:
 * MaxStack obj = new MaxStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.peekMax();
 * int param_5 = obj.popMax();
 */

File: Linkedin/LC716MaxStack.py
class MaxStack(object):

    def __init__(self):
        self.heap = []
        self.stack = []

    def push(self, x):
        node = [-x, -len(self.stack)]
        heapq.heappush(self.heap, node)
        self.stack.append(node)

    def pop(self):
        clean(self.heap, self.stack)
        node = self.stack.pop()
        node[1] = float('-inf')
        return -node[0]

    def top(self):
        clean(self.heap, self.stack)
        return -self.stack[-1][0]

    def peekMax(self):
        clean(self.heap, self.stack)
        return -self.heap[0][0]

    def popMax(self):
        clean(self.heap, self.stack)
        node = heapq.heappop(self.heap)
        node[1] = float('-inf')
        return -node[0]

def clean(heap, stack):
    while stack[-1][1] == float('-inf'):
        stack.pop()
    while heap[0][1] == float('-inf'):
        heapq.heappop(heap)


# Your MaxStack object will be instantiated and called as such:
# obj = MaxStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.peekMax()
# param_5 = obj.popMax()
